### ISSUE TITLE
Question about prob sparse attention implementation

### ISSUE THREAD

**Issue by cnut1648** [zhouhaoyi/Informer2020#257]
Hi, thanks for your awesome work!

I have some questions regarding how you implement the prob sparse attention, mainly in the `_prob_QK` function in the `ProbAttention` class.

1. you have the following
```
 # calculate the sampled Q_K
K_expand = K.unsqueeze(-3).expand(B, H, L_Q, L_K, E)
index_sample = torch.randint(L_K, (L_Q, sample_k)) # real U = U_part(factor*ln(L_k))*L_q
K_sample = K_expand[:, :, torch.arange(L_Q).unsqueeze(1), index_sample, :]
Q_K_sample = torch.matmul(Q.unsqueeze(-2), K_sample.transpose(-2, -1)).squeeze()
```
but I think the indexing doesn't work if `L_Q != L_K`, considering the following
```
B = H = 1
L_Q=3
L_K=4
E = 2
x = torch.arange(32).view(B, H,L_Q,L_K,E)
index = torch.randint(L_K, (L_Q, 2))
x[:, :, torch.arange(L_Q).unsqueeze(1), index, :]
```
this code would result in the error: ` shape '[1, 1, 3, 4, 2]' is invalid for input of size 32`

I think that the correct indexing might be this
```
random_idx = randint(L_x, (L_x, ))
K_sample = keys.index_select(2, random_idx)
Q_K_sample = matmul(Q, K_sample.transpose(-2, -1))
```
but this is doing indexing over dim=2, i.e. for each sample in the batch, for each head the indexing would be the same. I am not sure if this is what the authors want.

2. The code has 
```
M = Q_K_sample.max(-1)[0] - torch.div(Q_K_sample.sum(-1), L_K)
```
But why in the second term we have sum divided by L_K? I thought the last dimension is E? I am wondering if this should be
```
M =  Q_K_sample.max(-1)[0] - torch.mean(Q_K_sample, -1)
```

Thanks! Have a nice day!

**Comment by cookieminions** [comment]
1. We only use probsparse self-attention in Encoder's self-attention, so `L_Q==L_K`
2. It's OK to use `mean`
